# Análise de Vendas - Pedidos, Clientes e Reviews
Por: Matheus Mendonça Lopes

- Para esse projeto, serão analisadas três das tabelas fornecidas pelo *Kaggle*, que são:<br> 
    - `olist_orders_dataset.csv`, contendo os pedidos do *e-commerce*; 
    - `olist_customers_dataset.csv`: contendo os clientes da plataforma;
    - `olist_order_reviews_dataset.csv`: contendo os reviews de cada pedido;
<br>
Para cada tabela, será realizada uma análise exploratória inicial;
<br>
- Em seguida, será verificado:
    - Quais são os 20 clientes que mais realizaram pedidos;
    - Desses 20 clientes, quais deixam mais reviews com comentário que a média geral?
    - Desses 20 clientes, quais tem média de notas abaixo da média geral;
    - Dos 20 clientes, selecionar os que tiverem média de notas e comentários acima do geral.

## Importando DataFrames

In [350]:
import numpy as np
import pandas as pd

In [351]:
df_pedidos_raw = pd.read_csv('../data/olist_orders_dataset.csv')
df_pedidos_raw.head(10)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5  a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6  136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7  6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8  76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9  e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   
5    delivered      2017-07-09 21:57:05  2017-07-09 22:10:13   
6     invoiced      2017-04-11 12:22:08  2017-04-13 13:25:17   
7    delivered      2017-05-16 13:10:30  2017-05-16 13:22:11   
8    delivered      2017-01-23 18:29:09  2017-01-25 02:50:47   
9    delivered      2017-07-29 11:55:02  2017-07-29 12:05:32   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   
5          2017-07-11 14:58:04           2017-07-26 10:57:55   
6                          NaN                           NaN   
7          2017-05-22 10:07:46           2017-05-26 12:55:51   
8          2017-01-26 14:16:31           2017-02-02 14:08:10   
9          2017-08-10 19:45:24           2017-08-16 17:14:30   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00  
5           2017-08-01 00:00:00  
6           2017-05-09 00:00:00  
7           2017-06-07 00:00:00  
8           2017-03-06 00:00:00  
9           2017-08-23 00:00:00

In [352]:
df_clientes_raw = pd.read_csv('../data/olist_customers_dataset.csv')
df_clientes_raw.head(10)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4   
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177   
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f   
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d   
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  
5                     89254         jaragua do sul             SC  
6                      4534              sao paulo             SP  
7                     35182                timoteo             MG  
8                     81560               curitiba             PR  
9                     30575         belo horizonte             MG

In [353]:
df_reviews_pedidos_raw = pd.read_csv('../data/olist_order_reviews_dataset.csv')
df_reviews_pedidos_raw.head(10)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   
7  7c6400515c67679fbee952a7525281ef  c31a859e34e3adac22f376954e19b39d   
8  a3f6f7f6f433de0aefbb97da197c554c  9c214ac970e84273583ab523dfafd09b   
9  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   
5             1                  NaN   
6             5                  NaN   
7             5                  NaN   
8             5                  NaN   
9             4            recomendo   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
5                                                NaN  2018-04-13 00:00:00   
6                                                NaN  2017-07-16 00:00:00   
7                                                NaN  2018-08-14 00:00:00   
8                                                NaN  2017-05-17 00:00:00   
9  aparelho eficiente. no site a marca do aparelh...  2018-05-22 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53  
5     2018-04-16 00:39:37  
6     2017-07-18 19:30:34  
7     2018-08-14 21:36:06  
8     2017-05-18 12:05:37  
9     2018-05-23 16:45:47

## Análises Iniciais

#### Análise Inicial: Pedidos

In [354]:
df_pedidos_raw.shape

(99441, 8)

Podemos observar que temos uma tabela de 99441 linhas e 8 colunas. Os próximos passos da análise inicial devem ser:
    - Verificar quais são as colunas;
    - Verificar se a coluna de `ids` possui `NaN` ou alguma duplicata;
    - Trabalhar com um novo dataset sem colunas que não pertencem ao escopo.


In [355]:
df_pedidos_raw.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')

In [356]:
df_pedidos_raw['order_id'].nunique()

99441

In [357]:
df_pedidos_raw['order_id'].isna().sum()

0

Pode-se perceber que não há `NaN` na coluna de `order_id`, e todos seus valores são únicos. Portanto, se trata da chave primária de pedidos.<br>
Agora, as colunas ``order_purchase_timestamp``, ``order_approved_at``, ``order_delivered_carrier_date``, ``order_delivered_customer_date`` e ``order_estimated_delivery_date`` serão eliminadas para facilitar a leitura, considerando que elas não fazem parte do escopo da análise desejada.

In [358]:
df_pedidos = df_pedidos_raw[['order_id', 'customer_id', 'order_status']]
df_pedidos.head(10)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5  a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6  136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7  6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8  76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9  e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   

  order_status  
0    delivered  
1    delivered  
2    delivered  
3    delivered  
4    delivered  
5    delivered  
6     invoiced  
7    delivered  
8    delivered  
9    delivered

É possível identificar que `order_status` já possui valores diferentes pelo `.head()` acima. Portanto, será usado o `.value_counts()` para descobrir quantos valores de cada tipo existem.

In [359]:
df_pedidos['order_status'].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [360]:
df_pedidos['order_status'].isna().sum()

0

Para finalizar a análise inicial, será averiguado se, em `customer_id`, há valores repetidos e `NaN`.

In [361]:
df_pedidos['customer_id'].nunique()

99441

In [362]:
df_pedidos['customer_id'].isna().sum()

0

In [363]:
df_pedidos['customer_id'].dtype

dtype('O')

Portanto, as informações relevantes obtidas sobre o DataFrame `df_pedidos`, até o momento, foram:
- Núm. de Linhas: 99441
- Núm. de Colunas (sendo utilizadas): 3
- Colunas:
    * `order_id`:
        - 99441 valores únicos
        - Nenhum NaN
        - Chave primária
    * `customer_id`: 
        - 99441 valores únicos
        - Nenhum NaN
    * `order_status`:
        - 8 tipos de valores
        - Nenhum NaN

#### Análise Inicial: Clientes

In [364]:
df_clientes_raw.head(10)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4   
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177   
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f   
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d   
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP  
5                     89254         jaragua do sul             SC  
6                      4534              sao paulo             SP  
7                     35182                timoteo             MG  
8                     81560               curitiba             PR  
9                     30575         belo horizonte             MG

A localização do cliente não é pertinente para a análise do projeto, então a tabela será filtrada para conter apenas as colunas `customer_id` e `customer_unique_id`.

In [365]:
df_clientes = df_clientes_raw[['customer_id', 'customer_unique_id']]
df_clientes.head(10)

customer_id                customer_unique_id
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22

In [366]:
df_clientes.shape

(99441, 2)

Analisando a coluna `customer_unique_id`, verificando quantidade de valores únicos e `NaN`.

In [367]:
df_clientes['customer_unique_id'].nunique()

96096

In [368]:
df_clientes['customer_unique_id'].isna().sum()

0

A quantidade de valores únicos em `customer_unique_id` é menor que o número de linhas, e não há `NaN`; portanto, há valores repetidos.

Verificando informações da coluna `customer_id`. De acordo com a documentação dos datasets, `customer_id` se trata de um id único criado para cada pedido na tabela de pedidos, portanto, é uma chave estrangeira. Por conta disso, um cliente que faz mais de um pedido acaba tendo o `customer_unique_id` repetido na tabela de clientes.
<br>
Com essa informação, será possível descobrir quantos pedidos cada cliente realizou.

In [369]:
df_clientes['customer_id'].nunique()

99441

In [370]:
df_clientes['customer_id'].isna().sum()

0

In [371]:
df_clientes['customer_id'].dtype

dtype('O')

Portanto, as informações relevantes obtidas sobre o DataFrame `df_clientes`, até o momento, foram:
- Núm. de Linhas: 99441
- Núm. de Colunas (sendo utilizadas): 2
- Colunas:
    * `customer_id`: 
        - 99441 valores únicos
        - Nenhum NaN
        - Chave estrangeira proveniente de `df_pedidos`
    * `customer_unique_id`:
        - 96096 valores únicos
        - Nenhum NaN
        - Se trata do `id` pra cada cliente

#### Análise Inicial: Reviews 

In [372]:
df_reviews_pedidos_raw.head(10)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   
7  7c6400515c67679fbee952a7525281ef  c31a859e34e3adac22f376954e19b39d   
8  a3f6f7f6f433de0aefbb97da197c554c  9c214ac970e84273583ab523dfafd09b   
9  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   
5             1                  NaN   
6             5                  NaN   
7             5                  NaN   
8             5                  NaN   
9             4            recomendo   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   
5                                                NaN  2018-04-13 00:00:00   
6                                                NaN  2017-07-16 00:00:00   
7                                                NaN  2018-08-14 00:00:00   
8                                                NaN  2017-05-17 00:00:00   
9  aparelho eficiente. no site a marca do aparelh...  2018-05-22 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53  
5     2018-04-16 00:39:37  
6     2017-07-18 19:30:34  
7     2018-08-14 21:36:06  
8     2017-05-18 12:05:37  
9     2018-05-23 16:45:47

Considerando que datas não são importantes para o escopo da análise, as colunas `review_comment_title`, `review_creation_date` e `review_answer_timestamp` não serão utilizadas.

In [373]:
df_reviews = df_reviews_pedidos_raw[['review_id','order_id','review_score','review_comment_message']]
df_reviews.head(10)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   
7  7c6400515c67679fbee952a7525281ef  c31a859e34e3adac22f376954e19b39d   
8  a3f6f7f6f433de0aefbb97da197c554c  9c214ac970e84273583ab523dfafd09b   
9  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   

   review_score                             review_comment_message  
0             4                                                NaN  
1             5                                                NaN  
2             5                                                NaN  
3             5              Recebi bem antes do prazo estipulado.  
4             5  Parabéns lojas lannister adorei comprar pela I...  
5             1                                                NaN  
6             5                                                NaN  
7             5                                                NaN  
8             5                                                NaN  
9             4  aparelho eficiente. no site a marca do aparelh...

In [374]:
df_reviews.shape

(99224, 4)

O número de linhas é maior que o número de pedidos entregues. Isso pode se dar pois um mesmo pedido pode receber mais de uma review do cliente, ou uma mesma review pode ser atribuída para vários pedidos. As próximas análises abordarão isso.

Analisando as colunas `review_id` e `order_id`:

In [375]:
df_reviews['review_id'].nunique()

98410

In [376]:
df_reviews['review_id'].isna().sum()

0

In [377]:
df_reviews['order_id'].nunique()

98673

In [378]:
df_reviews['order_id'].isna().sum()

0

É possível que nenhuma das duas colunas possui `NaN`, mas existem valores em cada uma das colunas que se repete.<br>
Verificando o valor mais comum de cada uma delas e como aparecem na tabela:

In [379]:
most_common_review_id = df_reviews['review_id'].value_counts().nlargest(1).index[0]
most_common_review_id

'7b606b0d57b078384f0b58eac1d41d78'

In [380]:
df_reviews[df_reviews['review_id']==most_common_review_id]

review_id                          order_id  \
7500   7b606b0d57b078384f0b58eac1d41d78  f3028a8f41ea1ee2b461420913663f97   
59859  7b606b0d57b078384f0b58eac1d41d78  2deb17060fc1ce18a85eba953ddcdeaf   
61069  7b606b0d57b078384f0b58eac1d41d78  2f8f31eb2f7b6572836d662a6625c8e4   

       review_score review_comment_message  
7500              5                    NaN  
59859             5                    NaN  
61069             5                    NaN

Uma mesmo review pode ser utilizada pelo cliente para diferentes pedidos.

In [381]:
most_common_order_id = df_reviews['order_id'].value_counts().nlargest(1).index[0]
most_common_order_id

'c88b1d1b157a9999ce368f218a407141'

In [382]:
df_reviews[df_reviews['order_id']==most_common_order_id]

review_id                          order_id  \
1985   ffb8cff872a625632ac983eb1f88843c  c88b1d1b157a9999ce368f218a407141   
82525  202b5f44d09cd3cfc0d6bd12f01b044c  c88b1d1b157a9999ce368f218a407141   
89360  fb96ea2ef8cce1c888f4d45c8e22b793  c88b1d1b157a9999ce368f218a407141   

       review_score review_comment_message  
1985              3                    NaN  
82525             5                    NaN  
89360             5                    NaN

O usuário pode deixar diferentes reviews para um mesmo produto.

Analisando a coluna de `review_score`:

In [383]:
df_reviews['review_score'].value_counts(True)*100

review_score
5    57.776344
4    19.291704
1    11.513344
3     8.242965
2     3.175643
Name: proportion, dtype: float64

In [384]:
df_reviews['review_score'].describe()

count    99224.000000
mean         4.086421
std          1.347579
min          1.000000
25%          4.000000
50%          5.000000
75%          5.000000
max          5.000000
Name: review_score, dtype: float64

Percebe-se que a nota média de avaliações é muito próxima de 4, ainda que mais de 50% das avaliações seja 5.

Verificando a porcentagem de `NaN` na coluna `review_comment_message`.

In [385]:
nan_comment_percentage = df_reviews['review_comment_message'].isna().mean()*100
print(f"A porcetagem de NaN é: {nan_comment_percentage:.2f}%")

A porcetagem de NaN é: 58.70%


Portanto, as informações relevantes obtidas sobre o DataFrame `df_reviews`, até o momento, foram:
- Núm. de Linhas: 99224
- Núm. de Colunas (sendo utilizadas): 4
- Colunas:
    * `review_id`: 
        - 98410 valores únicos
        - Nenhum NaN
    * `order_id`:
        - 98673 valores únicos
        - Nenhum NaN
        - Forma chave composta com `review_id`
    * `review_score`:
        - Intervalo discreto de 1 a 5
        - Nenhum NaN
        - Média: 4,08
        - Porcetagens de cada valor
        
            |  review_score    |porcentagem|
            |:----------------:|:---------:|
            |5        |57,77%              |
            |4        |19,29%              |
            |3        |8,24%               |
            |2        |3,17%               |
            |1        |11,51%              |
    
    * `review_comment_message`:
        - 58,7% são NaN


## 20 Clientes que mais realizam pedidos

Para essa análise, precisaremos juntar os `customer_id` da tabela de pedidos e clientes e, com um `group_by`, descobrir quantos pedidos cada cliente fez.

In [386]:
df_merge_pedidos_clientes = pd.merge(df_clientes, df_pedidos, on='customer_id')
df_merge_pedidos_clientes

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

                               order_id order_status  
0      00e7ee1b050b8499577073aeb2a297a1    delivered  
1      29150127e6685892b6eab3eec79f59c7    delivered  
2      b2059ed67ce144a36e2aa97d2c9e9ad2    delivered  
3      951670f92359f4fe4a63112aa7306eba    delivered  
4      6b7d50bd145f6fc7f33cebabd7e49d0f    delivered  
...                                 ...          ...  
99436  6760e20addcf0121e9d58f2f1ff14298    delivered  
99437  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered  
99438  fed4434add09a6f332ea398efd656a5c    delivered  
99439  e31ec91cea1ecf97797787471f98a8c2    delivered  
99440  28db69209a75e59f20ccbb5c36a20b90    delivered  

[99441 rows x 4 columns]

In [387]:
df_num_pedidos_por_cliente = df_merge_pedidos_clientes.groupby('customer_unique_id')['order_id'].count().reset_index()
df_num_pedidos_por_cliente


customer_unique_id  order_id
0      0000366f3b9a7992bf8c76cfdf3221e2         1
1      0000b849f77a49e4a4ce2b2a4ca5be3f         1
2      0000f46a3911fa3c0805444483337064         1
3      0000f6ccb0745a6a4b88665a16c9f078         1
4      0004aac84e0df4da2b147fca70cf8255         1
...                                 ...       ...
96091  fffcf5a5ff07b0908bd4e2dbc735a684         1
96092  fffea47cd6d3cc0a88bd621562a9d061         1
96093  ffff371b4d645b6ecea244b27531430a         1
96094  ffff5962728ec6157033ef9805bacc48         1
96095  ffffd2657e2aad2907e67c3e9daecbeb         1

[96096 rows x 2 columns]

A tabela está ordenada pelo `customer_unique_id`, agora será modificada para estar ordenada pelo `order_id`, e essa coluna será renomeada para `order_count`.

In [388]:
df_num_pedidos_por_cliente.sort_values('order_id', ascending=False, inplace=True)
df_num_pedidos_por_cliente.rename(columns={'order_id': 'order_count'}, inplace=True)
df_num_pedidos_por_cliente

customer_unique_id  order_count
52973  8d50f5eadf50201ccdcedfb9e2ac8455           17
23472  3e43e6105506432c953e165fb2acf44c            9
37797  6469f99c1f9dfae7733b25662e7f1782            7
76082  ca77025e7201e3b30c44b472ff346268            7
10354  1b6c7548a2a1f9037c1fd3ddfed95f33            7
...                                 ...          ...
32508  5657dfebff5868c4dc7e8355fea865c4            1
32507  5657596addb4d7b07b32cd330614bdf8            1
32506  5656eb169546146caeab56c3ffc3d268            1
32505  5656a8fabc8629ff96b2bc14f8c09a27            1
96095  ffffd2657e2aad2907e67c3e9daecbeb            1

[96096 rows x 2 columns]

Tendo a tabela da quantidade de pedidos por cliente, podemos obter a tabela com os top 20 clientes.

In [389]:
df_top_20_clientes = df_num_pedidos_por_cliente.head(20)
df_top_20_clientes

customer_unique_id  order_count
52973  8d50f5eadf50201ccdcedfb9e2ac8455           17
23472  3e43e6105506432c953e165fb2acf44c            9
37797  6469f99c1f9dfae7733b25662e7f1782            7
76082  ca77025e7201e3b30c44b472ff346268            7
10354  1b6c7548a2a1f9037c1fd3ddfed95f33            7
7175   12f5d6e1cbf93dafd9dcc19095df0b3d            6
83540  de34b16117594161a6a89c50b289d35a            6
37585  63cfc61cee11cbe306bff5857d00bfe4            6
90451  f0e310a6839dce9de1638e0fe5ab282a            6
27043  47c1a3033b8b77b3ab6e109eb4d5fdf3            6
82883  dc813062e0fc23409cd255f7f53c7074            6
29512  4e65032f1f574189fb793bac5a867bbc            5
68007  b4e4f24de1e8725b74e4a1f4975116ed            5
21578  394ac4de8f3acb14253c177f0e15bc58            5
35628  5e8f38a9a1c023f3db718edcf926a2db            5
20337  35ecdf6858edc6427223b64804cf028e            5
43934  74cb1ad7e6d5674325c1f99b5ea30d82            5
95534  fe81bb32c243a86b2f86fbf053fe6140            5
32664  56c8638e7c058b98aae6d74d2dd6ea23            5
45647  795c1622cf7a53d63d324e862349d01c            4

## Dos 20, quais deixam mais reviews com comentário que a média geral?

In [390]:
df_merge_pedidos_clientes_reviews = df_merge_pedidos_clientes.merge(df_reviews, on='order_id')
df_merge_pedidos_clientes_reviews

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99219  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99220  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99221  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99222  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99223  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

                               order_id order_status  \
0      00e7ee1b050b8499577073aeb2a297a1    delivered   
1      29150127e6685892b6eab3eec79f59c7    delivered   
2      b2059ed67ce144a36e2aa97d2c9e9ad2    delivered   
3      951670f92359f4fe4a63112aa7306eba    delivered   
4      6b7d50bd145f6fc7f33cebabd7e49d0f    delivered   
...                                 ...          ...   
99219  6760e20addcf0121e9d58f2f1ff14298    delivered   
99220  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered   
99221  fed4434add09a6f332ea398efd656a5c    delivered   
99222  e31ec91cea1ecf97797787471f98a8c2    delivered   
99223  28db69209a75e59f20ccbb5c36a20b90    delivered   

                              review_id  review_score  \
0      88b8b52d46df026a9d1ad2136a59b30b             4   
1      02fc48a9efa3e3d0f1a8ea26507eeec3             5   
2      5ad6695d76ee186dc473c42706984d87             5   
3      059a801bb31f6aab2266e672cab87bc5             5   
4      8490879d58d6c5d7773f2739a03f089a             5   
...                                 ...           ...   
99219  36e2cdbaa9f639b57c53b37ac798fee8             4   
99220  b273b431c3aedb4eed18643309652940             5   
99221  fa4f16891e6b2edd1354668d07f5648b             1   
99222  0bcdc9e450ea500811a8d39ee993cd47             5   
99223  d95d701d062d2bb88041768a63aa750d             5   

                                  review_comment_message  
0                                                    NaN  
1                                                    NaN  
2                                                    NaN  
3                                                    NaN  
4          O baratheon è esxelente Amo adoro o baratheon  
...                                                  ...  
99219                                                NaN  
99220                                                NaN  
99221  Esperava qualidade no atendimento e estou tend...  
99222                                                NaN  
99223  Produto chegou bem antes do prazo , gostei mui...  

[99224 rows x 7 columns]

In [391]:
df_top_20_comentarios_sem_filtro = df_merge_pedidos_clientes_reviews.merge(df_top_20_clientes, on='customer_unique_id')
df_top_20_comentarios_sem_filtro

customer_id                customer_unique_id  \
0    a666bb7a77916be2716fa1420c73c56f  4e65032f1f574189fb793bac5a867bbc   
1    9551c9f70be99f31e88354c565e543f0  795c1622cf7a53d63d324e862349d01c   
2    db846bcd3ea0ef56effb856bd0266a90  394ac4de8f3acb14253c177f0e15bc58   
3    cfd34f15eb3c20730fcfa9da03936fc8  b4e4f24de1e8725b74e4a1f4975116ed   
4    cfd34f15eb3c20730fcfa9da03936fc8  b4e4f24de1e8725b74e4a1f4975116ed   
..                                ...                               ...   
127  9a1afef458843a022e431f4cb304dfe9  8d50f5eadf50201ccdcedfb9e2ac8455   
128  31dd055624c66f291578297a551a6cdf  8d50f5eadf50201ccdcedfb9e2ac8455   
129  ed1793d2d1e4175d5846ce7ebb4a01f5  12f5d6e1cbf93dafd9dcc19095df0b3d   
130  90e96fc65e2b0b97264512ed783f86e4  dc813062e0fc23409cd255f7f53c7074   
131  a969b9f8d923bc7fd97b578f7c499194  12f5d6e1cbf93dafd9dcc19095df0b3d   

                             order_id order_status  \
0    0136390286be8a34efdc801ed150568f    delivered   
1    58a1e508f021717acbcf12f9f328f497    delivered   
2    4cc8845d2a687fdd3d47898857cecdfd    delivered   
3    9eb9610a7e3c04e7ab0298d704c383dc    delivered   
4    9eb9610a7e3c04e7ab0298d704c383dc    delivered   
..                                ...          ...   
127  89d9b111d2b990deb5f5f9769f92800b    delivered   
128  5837a2c844decae8a778657425f6d664  unavailable   
129  747996a66f5aa711deb8ae58f5ae46a0    delivered   
130  2f7686472526a14dea7833c54aea6a70    delivered   
131  8a784d47854e4cbc5562362393d504db    delivered   

                            review_id  review_score  \
0    de23c38fb03573b2fd75e8099d50fa06             5   
1    755e9dbaedb9a4eb45eba3d180361476             5   
2    8983c4ffad0709da7e4f754e73107f6f             5   
3    d8a302244615111a0399d88db9916fd2             4   
4    a46a4abee1cb1ec1cc2470db9ea73de9             4   
..                                ...           ...   
127  ceea1f7a0d8f9c6d139da6813f09344f             5   
128  fe4defbc0ab6116a1e5a93741fde4b1d             1   
129  e5d48386325f21f0360c77b19c77feb1             5   
130  c784c3dac1aeecb954186037e912c615             5   
131  3d1a9381854446c25e98012c64bb5460             5   

                                review_comment_message  order_count  
0                                           Recomendo             5  
1                                                  NaN            4  
2                                                  NaN            5  
3                                  muito bom a compara            5  
4      muito bonito preço regular podia ser mai barato            5  
..                                                 ...          ...  
127                                                NaN           17  
128                                                NaN           17  
129  Recebi bem antes do prazo informado e o produt...            6  
130                                                NaN            6  
131                                                NaN            6  

[132 rows x 8 columns]

In [392]:
df_top_20_reviews = df_top_20_comentarios_sem_filtro[['customer_unique_id', 'review_score', 'review_comment_message']]
df_top_20_reviews

customer_unique_id  review_score  \
0    4e65032f1f574189fb793bac5a867bbc             5   
1    795c1622cf7a53d63d324e862349d01c             5   
2    394ac4de8f3acb14253c177f0e15bc58             5   
3    b4e4f24de1e8725b74e4a1f4975116ed             4   
4    b4e4f24de1e8725b74e4a1f4975116ed             4   
..                                ...           ...   
127  8d50f5eadf50201ccdcedfb9e2ac8455             5   
128  8d50f5eadf50201ccdcedfb9e2ac8455             1   
129  12f5d6e1cbf93dafd9dcc19095df0b3d             5   
130  dc813062e0fc23409cd255f7f53c7074             5   
131  12f5d6e1cbf93dafd9dcc19095df0b3d             5   

                                review_comment_message  
0                                           Recomendo   
1                                                  NaN  
2                                                  NaN  
3                                  muito bom a compara  
4      muito bonito preço regular podia ser mai barato  
..                                                 ...  
127                                                NaN  
128                                                NaN  
129  Recebi bem antes do prazo informado e o produt...  
130                                                NaN  
131                                                NaN  

[132 rows x 3 columns]

In [393]:
df_top_20_reviews = df_top_20_reviews.fillna('no comments left by customer.')
df_top_20_reviews

customer_unique_id  review_score  \
0    4e65032f1f574189fb793bac5a867bbc             5   
1    795c1622cf7a53d63d324e862349d01c             5   
2    394ac4de8f3acb14253c177f0e15bc58             5   
3    b4e4f24de1e8725b74e4a1f4975116ed             4   
4    b4e4f24de1e8725b74e4a1f4975116ed             4   
..                                ...           ...   
127  8d50f5eadf50201ccdcedfb9e2ac8455             5   
128  8d50f5eadf50201ccdcedfb9e2ac8455             1   
129  12f5d6e1cbf93dafd9dcc19095df0b3d             5   
130  dc813062e0fc23409cd255f7f53c7074             5   
131  12f5d6e1cbf93dafd9dcc19095df0b3d             5   

                                review_comment_message  
0                                           Recomendo   
1                        no comments left by customer.  
2                        no comments left by customer.  
3                                  muito bom a compara  
4      muito bonito preço regular podia ser mai barato  
..                                                 ...  
127                      no comments left by customer.  
128                      no comments left by customer.  
129  Recebi bem antes do prazo informado e o produt...  
130                      no comments left by customer.  
131                      no comments left by customer.  

[132 rows x 3 columns]

In [394]:
top_20_reviews_count = df_top_20_reviews.groupby('customer_unique_id')['review_comment_message'].count().reset_index().sort_values('review_comment_message', ascending=False)
top_20_reviews_count

customer_unique_id  review_comment_message
13  8d50f5eadf50201ccdcedfb9e2ac8455                      17
4   3e43e6105506432c953e165fb2acf44c                       9
10  6469f99c1f9dfae7733b25662e7f1782                       7
5   47c1a3033b8b77b3ab6e109eb4d5fdf3                       7
15  ca77025e7201e3b30c44b472ff346268                       7
14  b4e4f24de1e8725b74e4a1f4975116ed                       7
1   1b6c7548a2a1f9037c1fd3ddfed95f33                       7
11  74cb1ad7e6d5674325c1f99b5ea30d82                       6
18  f0e310a6839dce9de1638e0fe5ab282a                       6
17  de34b16117594161a6a89c50b289d35a                       6
16  dc813062e0fc23409cd255f7f53c7074                       6
0   12f5d6e1cbf93dafd9dcc19095df0b3d                       6
9   63cfc61cee11cbe306bff5857d00bfe4                       6
2   35ecdf6858edc6427223b64804cf028e                       6
8   5e8f38a9a1c023f3db718edcf926a2db                       5
7   56c8638e7c058b98aae6d74d2dd6ea23                       5
6   4e65032f1f574189fb793bac5a867bbc                       5
3   394ac4de8f3acb14253c177f0e15bc58                       5
19  fe81bb32c243a86b2f86fbf053fe6140                       5
12  795c1622cf7a53d63d324e862349d01c                       4

In [395]:
df_top_20_info = pd.merge(df_top_20_clientes, top_20_reviews_count, on='customer_unique_id')
df_top_20_info.rename(columns={'review_comment_message':'review_count'}, inplace=True)
df_top_20_info

customer_unique_id  order_count  review_count
0   8d50f5eadf50201ccdcedfb9e2ac8455           17            17
1   3e43e6105506432c953e165fb2acf44c            9             9
2   6469f99c1f9dfae7733b25662e7f1782            7             7
3   ca77025e7201e3b30c44b472ff346268            7             7
4   1b6c7548a2a1f9037c1fd3ddfed95f33            7             7
5   12f5d6e1cbf93dafd9dcc19095df0b3d            6             6
6   de34b16117594161a6a89c50b289d35a            6             6
7   63cfc61cee11cbe306bff5857d00bfe4            6             6
8   f0e310a6839dce9de1638e0fe5ab282a            6             6
9   47c1a3033b8b77b3ab6e109eb4d5fdf3            6             7
10  dc813062e0fc23409cd255f7f53c7074            6             6
11  4e65032f1f574189fb793bac5a867bbc            5             5
12  b4e4f24de1e8725b74e4a1f4975116ed            5             7
13  394ac4de8f3acb14253c177f0e15bc58            5             5
14  5e8f38a9a1c023f3db718edcf926a2db            5             5
15  35ecdf6858edc6427223b64804cf028e            5             6
16  74cb1ad7e6d5674325c1f99b5ea30d82            5             6
17  fe81bb32c243a86b2f86fbf053fe6140            5             5
18  56c8638e7c058b98aae6d74d2dd6ea23            5             5
19  795c1622cf7a53d63d324e862349d01c            4             4

In [396]:
mask = df_top_20_reviews['review_comment_message'] == 'no comments left by customer.'
mask

0      False
1       True
2       True
3      False
4      False
       ...  
127     True
128     True
129    False
130     True
131     True
Name: review_comment_message, Length: 132, dtype: bool

In [397]:
top_20_nocomments_count = df_top_20_reviews[mask].groupby('customer_unique_id')['review_comment_message'].count().reset_index().sort_values('review_comment_message', ascending=False)
top_20_nocomments_count

customer_unique_id  review_comment_message
11  8d50f5eadf50201ccdcedfb9e2ac8455                      17
4   3e43e6105506432c953e165fb2acf44c                       9
8   6469f99c1f9dfae7733b25662e7f1782                       7
5   47c1a3033b8b77b3ab6e109eb4d5fdf3                       7
1   1b6c7548a2a1f9037c1fd3ddfed95f33                       7
12  ca77025e7201e3b30c44b472ff346268                       7
15  f0e310a6839dce9de1638e0fe5ab282a                       6
9   74cb1ad7e6d5674325c1f99b5ea30d82                       6
2   35ecdf6858edc6427223b64804cf028e                       5
3   394ac4de8f3acb14253c177f0e15bc58                       5
7   63cfc61cee11cbe306bff5857d00bfe4                       5
6   56c8638e7c058b98aae6d74d2dd6ea23                       4
10  795c1622cf7a53d63d324e862349d01c                       4
13  dc813062e0fc23409cd255f7f53c7074                       3
0   12f5d6e1cbf93dafd9dcc19095df0b3d                       2
16  fe81bb32c243a86b2f86fbf053fe6140                       2
14  de34b16117594161a6a89c50b289d35a                       1

In [398]:
df_top_20_info = df_top_20_info.merge(top_20_nocomments_count, on='customer_unique_id', how='left')
df_top_20_info.rename(columns={'review_comment_message':'no_comment_review_count'}, inplace=True)
df_top_20_info['no_comment_review_count'] = df_top_20_info['no_comment_review_count'].fillna(0).astype(np.int64)
df_top_20_info

customer_unique_id  order_count  review_count  \
0   8d50f5eadf50201ccdcedfb9e2ac8455           17            17   
1   3e43e6105506432c953e165fb2acf44c            9             9   
2   6469f99c1f9dfae7733b25662e7f1782            7             7   
3   ca77025e7201e3b30c44b472ff346268            7             7   
4   1b6c7548a2a1f9037c1fd3ddfed95f33            7             7   
5   12f5d6e1cbf93dafd9dcc19095df0b3d            6             6   
6   de34b16117594161a6a89c50b289d35a            6             6   
7   63cfc61cee11cbe306bff5857d00bfe4            6             6   
8   f0e310a6839dce9de1638e0fe5ab282a            6             6   
9   47c1a3033b8b77b3ab6e109eb4d5fdf3            6             7   
10  dc813062e0fc23409cd255f7f53c7074            6             6   
11  4e65032f1f574189fb793bac5a867bbc            5             5   
12  b4e4f24de1e8725b74e4a1f4975116ed            5             7   
13  394ac4de8f3acb14253c177f0e15bc58            5             5   
14  5e8f38a9a1c023f3db718edcf926a2db            5             5   
15  35ecdf6858edc6427223b64804cf028e            5             6   
16  74cb1ad7e6d5674325c1f99b5ea30d82            5             6   
17  fe81bb32c243a86b2f86fbf053fe6140            5             5   
18  56c8638e7c058b98aae6d74d2dd6ea23            5             5   
19  795c1622cf7a53d63d324e862349d01c            4             4   

    no_comment_review_count  
0                        17  
1                         9  
2                         7  
3                         7  
4                         7  
5                         2  
6                         1  
7                         5  
8                         6  
9                         7  
10                        3  
11                        0  
12                        0  
13                        5  
14                        0  
15                        5  
16                        6  
17                        2  
18                        4  
19                        4

In [399]:
no_comment_percentage = np.array((100*df_top_20_info['no_comment_review_count']/df_top_20_info['review_count']).round(2))
no_comment_percentage

array([100.  , 100.  , 100.  , 100.  , 100.  ,  33.33,  16.67,  83.33,
       100.  , 100.  ,  50.  ,   0.  ,   0.  , 100.  ,   0.  ,  83.33,
       100.  ,  40.  ,  80.  , 100.  ])

In [400]:
df_top_20_info['no_comment_percentage'] = no_comment_percentage
df_top_20_info

customer_unique_id  order_count  review_count  \
0   8d50f5eadf50201ccdcedfb9e2ac8455           17            17   
1   3e43e6105506432c953e165fb2acf44c            9             9   
2   6469f99c1f9dfae7733b25662e7f1782            7             7   
3   ca77025e7201e3b30c44b472ff346268            7             7   
4   1b6c7548a2a1f9037c1fd3ddfed95f33            7             7   
5   12f5d6e1cbf93dafd9dcc19095df0b3d            6             6   
6   de34b16117594161a6a89c50b289d35a            6             6   
7   63cfc61cee11cbe306bff5857d00bfe4            6             6   
8   f0e310a6839dce9de1638e0fe5ab282a            6             6   
9   47c1a3033b8b77b3ab6e109eb4d5fdf3            6             7   
10  dc813062e0fc23409cd255f7f53c7074            6             6   
11  4e65032f1f574189fb793bac5a867bbc            5             5   
12  b4e4f24de1e8725b74e4a1f4975116ed            5             7   
13  394ac4de8f3acb14253c177f0e15bc58            5             5   
14  5e8f38a9a1c023f3db718edcf926a2db            5             5   
15  35ecdf6858edc6427223b64804cf028e            5             6   
16  74cb1ad7e6d5674325c1f99b5ea30d82            5             6   
17  fe81bb32c243a86b2f86fbf053fe6140            5             5   
18  56c8638e7c058b98aae6d74d2dd6ea23            5             5   
19  795c1622cf7a53d63d324e862349d01c            4             4   

    no_comment_review_count  no_comment_percentage  
0                        17                 100.00  
1                         9                 100.00  
2                         7                 100.00  
3                         7                 100.00  
4                         7                 100.00  
5                         2                  33.33  
6                         1                  16.67  
7                         5                  83.33  
8                         6                 100.00  
9                         7                 100.00  
10                        3                  50.00  
11                        0                   0.00  
12                        0                   0.00  
13                        5                 100.00  
14                        0                   0.00  
15                        5                  83.33  
16                        6                 100.00  
17                        2                  40.00  
18                        4                  80.00  
19                        4                 100.00

In [401]:
porcentagem_comentarios_nan = 58.7
mask = df_top_20_info['no_comment_percentage']<porcentagem_comentarios_nan
df_top_20_info[mask]

customer_unique_id  order_count  review_count  \
5   12f5d6e1cbf93dafd9dcc19095df0b3d            6             6   
6   de34b16117594161a6a89c50b289d35a            6             6   
10  dc813062e0fc23409cd255f7f53c7074            6             6   
11  4e65032f1f574189fb793bac5a867bbc            5             5   
12  b4e4f24de1e8725b74e4a1f4975116ed            5             7   
14  5e8f38a9a1c023f3db718edcf926a2db            5             5   
17  fe81bb32c243a86b2f86fbf053fe6140            5             5   

    no_comment_review_count  no_comment_percentage  
5                         2                  33.33  
6                         1                  16.67  
10                        3                  50.00  
11                        0                   0.00  
12                        0                   0.00  
14                        0                   0.00  
17                        2                  40.00

## Dos 20, quais têm média de nota abaixo da média geral?

In [402]:
df_top_20_reviews

customer_unique_id  review_score  \
0    4e65032f1f574189fb793bac5a867bbc             5   
1    795c1622cf7a53d63d324e862349d01c             5   
2    394ac4de8f3acb14253c177f0e15bc58             5   
3    b4e4f24de1e8725b74e4a1f4975116ed             4   
4    b4e4f24de1e8725b74e4a1f4975116ed             4   
..                                ...           ...   
127  8d50f5eadf50201ccdcedfb9e2ac8455             5   
128  8d50f5eadf50201ccdcedfb9e2ac8455             1   
129  12f5d6e1cbf93dafd9dcc19095df0b3d             5   
130  dc813062e0fc23409cd255f7f53c7074             5   
131  12f5d6e1cbf93dafd9dcc19095df0b3d             5   

                                review_comment_message  
0                                           Recomendo   
1                        no comments left by customer.  
2                        no comments left by customer.  
3                                  muito bom a compara  
4      muito bonito preço regular podia ser mai barato  
..                                                 ...  
127                      no comments left by customer.  
128                      no comments left by customer.  
129  Recebi bem antes do prazo informado e o produt...  
130                      no comments left by customer.  
131                      no comments left by customer.  

[132 rows x 3 columns]

In [403]:
top_20_avg_score = df_top_20_reviews.groupby('customer_unique_id')['review_score'].mean().reset_index().sort_values('review_score', ascending=False)
top_20_avg_score.rename(columns={'review_score': 'avg_score'}, inplace=True)
top_20_avg_score

customer_unique_id  avg_score
0   12f5d6e1cbf93dafd9dcc19095df0b3d   5.000000
6   4e65032f1f574189fb793bac5a867bbc   5.000000
15  ca77025e7201e3b30c44b472ff346268   5.000000
11  74cb1ad7e6d5674325c1f99b5ea30d82   5.000000
1   1b6c7548a2a1f9037c1fd3ddfed95f33   5.000000
10  6469f99c1f9dfae7733b25662e7f1782   5.000000
3   394ac4de8f3acb14253c177f0e15bc58   5.000000
5   47c1a3033b8b77b3ab6e109eb4d5fdf3   4.857143
13  8d50f5eadf50201ccdcedfb9e2ac8455   4.764706
9   63cfc61cee11cbe306bff5857d00bfe4   4.500000
16  dc813062e0fc23409cd255f7f53c7074   4.500000
18  f0e310a6839dce9de1638e0fe5ab282a   4.500000
7   56c8638e7c058b98aae6d74d2dd6ea23   4.200000
19  fe81bb32c243a86b2f86fbf053fe6140   4.200000
14  b4e4f24de1e8725b74e4a1f4975116ed   4.142857
12  795c1622cf7a53d63d324e862349d01c   4.000000
2   35ecdf6858edc6427223b64804cf028e   4.000000
8   5e8f38a9a1c023f3db718edcf926a2db   3.800000
17  de34b16117594161a6a89c50b289d35a   3.000000
4   3e43e6105506432c953e165fb2acf44c   2.777778

In [404]:
df_top_20_info = df_top_20_info.merge(top_20_avg_score, on='customer_unique_id')
df_top_20_info

customer_unique_id  order_count  review_count  \
0   8d50f5eadf50201ccdcedfb9e2ac8455           17            17   
1   3e43e6105506432c953e165fb2acf44c            9             9   
2   6469f99c1f9dfae7733b25662e7f1782            7             7   
3   ca77025e7201e3b30c44b472ff346268            7             7   
4   1b6c7548a2a1f9037c1fd3ddfed95f33            7             7   
5   12f5d6e1cbf93dafd9dcc19095df0b3d            6             6   
6   de34b16117594161a6a89c50b289d35a            6             6   
7   63cfc61cee11cbe306bff5857d00bfe4            6             6   
8   f0e310a6839dce9de1638e0fe5ab282a            6             6   
9   47c1a3033b8b77b3ab6e109eb4d5fdf3            6             7   
10  dc813062e0fc23409cd255f7f53c7074            6             6   
11  4e65032f1f574189fb793bac5a867bbc            5             5   
12  b4e4f24de1e8725b74e4a1f4975116ed            5             7   
13  394ac4de8f3acb14253c177f0e15bc58            5             5   
14  5e8f38a9a1c023f3db718edcf926a2db            5             5   
15  35ecdf6858edc6427223b64804cf028e            5             6   
16  74cb1ad7e6d5674325c1f99b5ea30d82            5             6   
17  fe81bb32c243a86b2f86fbf053fe6140            5             5   
18  56c8638e7c058b98aae6d74d2dd6ea23            5             5   
19  795c1622cf7a53d63d324e862349d01c            4             4   

    no_comment_review_count  no_comment_percentage  avg_score  
0                        17                 100.00   4.764706  
1                         9                 100.00   2.777778  
2                         7                 100.00   5.000000  
3                         7                 100.00   5.000000  
4                         7                 100.00   5.000000  
5                         2                  33.33   5.000000  
6                         1                  16.67   3.000000  
7                         5                  83.33   4.500000  
8                         6                 100.00   4.500000  
9                         7                 100.00   4.857143  
10                        3                  50.00   4.500000  
11                        0                   0.00   5.000000  
12                        0                   0.00   4.142857  
13                        5                 100.00   5.000000  
14                        0                   0.00   3.800000  
15                        5                  83.33   4.000000  
16                        6                 100.00   5.000000  
17                        2                  40.00   4.200000  
18                        4                  80.00   4.200000  
19                        4                 100.00   4.000000

In [405]:
media_geral = 4.08
mask = df_top_20_info['avg_score']<media_geral
df_top_20_info[mask]

customer_unique_id  order_count  review_count  \
1   3e43e6105506432c953e165fb2acf44c            9             9   
6   de34b16117594161a6a89c50b289d35a            6             6   
14  5e8f38a9a1c023f3db718edcf926a2db            5             5   
15  35ecdf6858edc6427223b64804cf028e            5             6   
19  795c1622cf7a53d63d324e862349d01c            4             4   

    no_comment_review_count  no_comment_percentage  avg_score  
1                         9                 100.00   2.777778  
6                         1                  16.67   3.000000  
14                        0                   0.00   3.800000  
15                        5                  83.33   4.000000  
19                        4                 100.00   4.000000

## Clientes especiais: avaliações e comentários acima da média geral

In [410]:
cliente_especial = np.array((df_top_20_info['avg_score']>media_geral) & (df_top_20_info['no_comment_percentage']<porcentagem_comentarios_nan))
cliente_especial

array([False, False, False, False, False,  True, False, False, False,
       False,  True,  True,  True, False, False, False, False,  True,
       False, False])

In [411]:
df_top_20_info['special_client'] = cliente_especial
df_top_20_info

customer_unique_id  order_count  review_count  \
0   8d50f5eadf50201ccdcedfb9e2ac8455           17            17   
1   3e43e6105506432c953e165fb2acf44c            9             9   
2   6469f99c1f9dfae7733b25662e7f1782            7             7   
3   ca77025e7201e3b30c44b472ff346268            7             7   
4   1b6c7548a2a1f9037c1fd3ddfed95f33            7             7   
5   12f5d6e1cbf93dafd9dcc19095df0b3d            6             6   
6   de34b16117594161a6a89c50b289d35a            6             6   
7   63cfc61cee11cbe306bff5857d00bfe4            6             6   
8   f0e310a6839dce9de1638e0fe5ab282a            6             6   
9   47c1a3033b8b77b3ab6e109eb4d5fdf3            6             7   
10  dc813062e0fc23409cd255f7f53c7074            6             6   
11  4e65032f1f574189fb793bac5a867bbc            5             5   
12  b4e4f24de1e8725b74e4a1f4975116ed            5             7   
13  394ac4de8f3acb14253c177f0e15bc58            5             5   
14  5e8f38a9a1c023f3db718edcf926a2db            5             5   
15  35ecdf6858edc6427223b64804cf028e            5             6   
16  74cb1ad7e6d5674325c1f99b5ea30d82            5             6   
17  fe81bb32c243a86b2f86fbf053fe6140            5             5   
18  56c8638e7c058b98aae6d74d2dd6ea23            5             5   
19  795c1622cf7a53d63d324e862349d01c            4             4   

    no_comment_review_count  no_comment_percentage  avg_score  special_client  
0                        17                 100.00   4.764706           False  
1                         9                 100.00   2.777778           False  
2                         7                 100.00   5.000000           False  
3                         7                 100.00   5.000000           False  
4                         7                 100.00   5.000000           False  
5                         2                  33.33   5.000000            True  
6                         1                  16.67   3.000000           False  
7                         5                  83.33   4.500000           False  
8                         6                 100.00   4.500000           False  
9                         7                 100.00   4.857143           False  
10                        3                  50.00   4.500000            True  
11                        0                   0.00   5.000000            True  
12                        0                   0.00   4.142857            True  
13                        5                 100.00   5.000000           False  
14                        0                   0.00   3.800000           False  
15                        5                  83.33   4.000000           False  
16                        6                 100.00   5.000000           False  
17                        2                  40.00   4.200000            True  
18                        4                  80.00   4.200000           False  
19                        4                 100.00   4.000000           False